sdk/python/feast/infra/offline_stores/remote.py

```diff
@@ -0,0 +1,401 @@
+    entity_schema = _get_entity_schema(
+        entity_df=entity_df,
+    )
+
+    event_timestamp_col = offline_utils.infer_event_timestamp_from_entity_df(
+        entity_schema=entity_schema,
+    )
+
+    timestamp_range = _get_entity_df_event_timestamp_range(
+        entity_df, event_timestamp_col
+    )
+
+    return RetrievalMetadata(
+        features=feature_refs,
+        keys=list(set(entity_df.columns) - {event_timestamp_col}),
+        min_event_timestamp=timestamp_range[0],
+        max_event_timestamp=timestamp_range[1],
+    )
+
+
+def _get_entity_schema(entity_df: pd.DataFrame) -> Dict[str, np.dtype]:
+    return dict(zip(entity_df.columns, entity_df.dtypes))
+
+
+def _get_entity_df_event_timestamp_range(
+    entity_df: Union[pd.DataFrame, str],
+    entity_df_event_timestamp_col: str,
+) -> Tuple[datetime, datetime]:
+    if not isinstance(entity_df, pd.DataFrame):
+        raise ValueError(
+            f"Please provide an entity_df of type {type(pd.DataFrame)} instead of type {type(entity_df)}"
+        )
+
+    entity_df_event_timestamp = entity_df.loc[
+        :, entity_df_event_timestamp_col
+    ].infer_objects()
+    if pd.api.types.is_string_dtype(entity_df_event_timestamp):
+        entity_df_event_timestamp = pd.to_datetime(entity_df_event_timestamp, utc=True)
+
+    return (
+        entity_df_event_timestamp.min().to_pydatetime(),
+        entity_df_event_timestamp.max().to_pydatetime(),
+    )
+
+
+def _send_retrieve_remote(
+    api: str,
+    api_parameters: Dict[str, Any],
+    entity_df: Union[pd.DataFrame, str],
+    table: pa.Table,
+    client: fl.FlightClient,
+):
+    command_descriptor = _call_put(api, api_parameters, client, entity_df, table)
+    return _call_get(client, command_descriptor)
+
+
+def _call_get(client, command_descriptor):
+    flight = client.get_flight_info(command_descriptor)
+    ticket = flight.endpoints[0].ticket
+    reader = client.do_get(ticket)
+    return reader.read_all()
+
+
+def _call_put(api, api_parameters, client, entity_df, table):
+    # Generate unique command identifier
+    command_id = str(uuid.uuid4())
+    command = {
+        "command_id": command_id,
+        "api": api,
+    }
+    # Add api parameters to command
+    for key, value in api_parameters.items():
+        command[key] = value
+
+    command_descriptor = fl.FlightDescriptor.for_command(
+        json.dumps(
+            command,
+        )
+    )
+
+    _put_parameters(command_descriptor, entity_df, table, client)
+    return command_descriptor
+
+
+def _put_parameters(
+    command_descriptor,
+    entity_df: Union[pd.DataFrame, str],
+    table: pa.Table,
+    client: fl.FlightClient,
+):
+    updatedTable: pa.Table
+
+    if entity_df is not None:
+        updatedTable = pa.Table.from_pandas(entity_df)
+    elif table is not None:
+        updatedTable = table
+    else:
+        updatedTable = _create_empty_table()
```

Comment:
@tmihalac I think this is no longer necessary after your changes in `persist`. `_put_parameters` is always called with either `entity_df` or `table` set.